File: 子串/76.最小覆盖子串.go
package 子串

func minWindow(s string, t string) string {
	nt := len(t)
	ns := len(s)
	if nt == 0 || ns < nt {
		return ""
	}
	need := make(map[byte]int, nt)
	for i := 0; i < len(t); i++ {
		need[t[i]]++
	}
	win := make(map[byte]int, nt)
	valid, left, start, minLen := 0, 0, 0, ns+1
	for right := 0; right < len(s); right++ {
		v := s[right]
		if _, ok := need[v]; ok {
			win[v]++
			if win[v] == need[v] {
				valid++
			}
		}
		for valid == len(need) {
			if right-left+1 < minLen {
				minLen = right - left + 1
				start = left
			}
			d := s[left]
			left++
			if _, ok := need[d]; ok {
				if win[d] == need[d] {
					valid--
				}
				win[d]--
			}
		}
	}
	if minLen == ns+1 {
		return ""
	}
	return s[start : start+minLen]
}

File: 子串/76.最小覆盖子串.java
class Solution {
    public String minWindow(String s, String t) {
        if(t.isEmpty() || s.length()<t.length()) return "";
        Map<Character,Integer> need = new HashMap<>();
        for(char ch : t.toCharArray()){
            need.put(ch,need.getOrDefault(ch,0)+1);
        }
        Map<Character,Integer> win = new HashMap<>();
        int valid=0,left=0,start=0,minLen= Integer.MAX_VALUE;

        for(int right=0;right<s.length();right++){
            char c = s.charAt(right);
            if(need.containsKey(c)){
                win.put(c,win.getOrDefault(c,0)+1);
                if(win.get(c).intValue() == need.get(c).intValue()) valid++;
            }
            while(valid == need.size()){
                if(right-left+1 < minLen){
                    minLen=right-left+1;
                    start=left;
                }
                char d = s.charAt(left++);
                if(need.containsKey(d)){
                    if(win.get(d).intValue() == need.get(d).intValue()) valid--;
                    win.put(d,win.get(d)-1);
                }
            }
        }
        return minLen == Integer.MAX_VALUE? "":s.substring(start,start+minLen);
    }
}